File: day_5/part_2.py
def get_inp(inp):
    ranges = []
    with open(inp, "r") as file:
        lines = file.readlines()
        for line in lines:
            line = line.strip()
            if not len(line):
                break
            ranges.append(list(map(lambda x: int(x), line.split("-"))))
    return ranges


def solve(inp):
    ranges = get_inp(inp)
    ranges.sort()
    start, end = -1, -1
    tmp = []
    for s2, e2 in ranges:
        if s2 > end:
            tmp.append([start, end])
            start = s2
            end = e2
        else:
            end = max(end, e2)
    tmp.append([start, end])
    res = 0
    for i in range(1, len(tmp)):
        res += tmp[i][1] - tmp[i][0] + 1

    return res


# print(solve("example.txt"))
print(solve("input.txt"))

File: day_5/part2.js
const fs = require("fs");
const FILEPATH = "./input.txt";
const filecontent = fs.readFileSync(FILEPATH, "utf-8").split("\n");

const solve = (inp) => {
  let ranges = inp.slice(0, inp.indexOf(""));
  ranges = ranges
    .map((x) => x.split("-").map((x) => +x))
    .sort((x, y) => x[0] - y[0]);
  let res = [];
  let i = 1;
  let [tmpSmall, tmpLarge] = [ranges[0][0], ranges[0][1]];
  while (i < ranges.length) {
    let [p1, p2] = ranges[i];
    if (p1 > tmpLarge) {
      res.push([tmpSmall, tmpLarge]);
      tmpSmall = p1;
      tmpLarge = p2;
    } else {
      tmpLarge = Math.max(p2, tmpLarge);
    }
    i += 1;
  }
  res.push([tmpSmall, tmpLarge]);
  return res.reduce((acum, cur) => acum + cur[1] - cur[0] + 1, 0);
};

console.log(solve(filecontent));
